recipes/wpewebkit/0013-Use-ASharedMemory-instead-of-shm.patch

```diff
@@ -0,0 +1,126 @@
+From 8e2319064b9d44394df5a52695c4a691d2f422cb Mon Sep 17 00:00:00 2001
+From: Jani Hautakangas <jani@igalia.com>
+Date: Wed, 28 Sep 2022 15:13:08 +0300
+Subject: [PATCH] Use ASharedMemory instead of shm
+
+---
+ .../WebKit/Platform/unix/SharedMemoryUnix.cpp | 60 ++++++++++---------
+ 1 file changed, 31 insertions(+), 29 deletions(-)
+
+diff --git a/Source/WebKit/Platform/unix/SharedMemoryUnix.cpp b/Source/WebKit/Platform/unix/SharedMemoryUnix.cpp
+index 998a2a76..11587009 100644
+--- a/Source/WebKit/Platform/unix/SharedMemoryUnix.cpp
++++ b/Source/WebKit/Platform/unix/SharedMemoryUnix.cpp
+@@ -54,6 +54,8 @@
+ #include "ArgumentCoders.h"
+ #endif
+ 
++#include <android/sharedmem.h>
++
+ namespace WebKit {
+ 
+ SharedMemory::Handle::Handle()
+@@ -124,56 +126,35 @@ static inline int accessModeMMap(SharedMemory::Protection protection)
+     return PROT_READ | PROT_WRITE;
+ }
+ 
+-static int createSharedMemory()
++static int createSharedMemory(size_t size)
+ {
+     int fileDescriptor = -1;
+ 
+-#if HAVE(LINUX_MEMFD_H)
+-    static bool isMemFdAvailable = true;
+-    if (isMemFdAvailable) {
+-        do {
+-            fileDescriptor = syscall(__NR_memfd_create, "WebKitSharedMemory", MFD_CLOEXEC);
+-        } while (fileDescriptor == -1 && errno == EINTR);
+-
+-        if (fileDescriptor != -1)
+-            return fileDescriptor;
+-
+-        if (errno != ENOSYS)
+-            return fileDescriptor;
+-
+-        isMemFdAvailable = false;
+-    }
+-#endif
+-
+     CString tempName;
+     for (int tries = 0; fileDescriptor == -1 && tries < 10; ++tries) {
+         String name = String("/WK2SharedMemory.") + String::number(static_cast<unsigned>(WTF::randomNumber() * (std::numeric_limits<unsigned>::max() + 1.0)));
+         tempName = name.utf8();
+ 
+         do {
+-            fileDescriptor = shm_open(tempName.data(), O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
++            fileDescriptor = ASharedMemory_create(tempName.data(), size);
+         } while (fileDescriptor == -1 && errno == EINTR);
+     }
+ 
+-    if (fileDescriptor != -1)
+-        shm_unlink(tempName.data());
+-
+     return fileDescriptor;
+ }
+ 
+ RefPtr<SharedMemory> SharedMemory::allocate(size_t size)
+ {
+-    int fileDescriptor = createSharedMemory();
++    int fileDescriptor = createSharedMemory(size);
+     if (fileDescriptor == -1) {
+         WTFLogAlways("Failed to create shared memory: %s", strerror(errno));
+         return nullptr;
+     }
+ 
+-    while (ftruncate(fileDescriptor, size) == -1) {
+-        if (errno != EINTR) {
+-            closeWithRetry(fileDescriptor);
+-            return nullptr;
+-        }
++    int err = ASharedMemory_setProt(fileDescriptor, PROT_READ | PROT_WRITE | PROT_EXEC);
```

Comment:
According [to the documentation](https://developer.android.com/ndk/reference/group/memory#group___memory_1ga5bbfa70e48fcbc488a8cb8fa7657c878) the default is already `PROT_READ | PROT_WRITE | PROT_EXEC`. If something, I think this should be changed to `PROT_READ | PROT_WRITE` because only memory used for JIT'd code needs to be executable, and allowing execution by default can result in arbitrary code execution security issues. Note that memory used by the JSC JIT is allocated through bmalloc so this function in theory should never be used by WebKit to allocate executable memory pages.

Reply:
good point. Removed PROT_EXEC